genie-web/src/main/java/com/netflix/genie/web/tasks/leader/ClusterCheckerTask.java

```diff
@@ -109,70 +121,99 @@ public ClusterCheckerTask(
-                    try {
-                        restTemplate.getForObject(this.scheme + host + this.healthEndpoint, String.class);
-                    } catch (final Exception ioe) {
-                        log.error("Unable to reach {}", host, ioe);
-                        badNodes.add(host);
-                    }
+            .forEach(this::validateHostAndUpdateErrorCount);
+
+        this.errorCounts.entrySet().removeIf(entry -> {
+            final String host = entry.getKey();
+            boolean result = true;
+            if (entry.getValue() >= properties.getLostThreshold()) {
+                try {
+                    updateJobsToFailedOnHost(host);
+                } catch (Exception e) {
+                    log.error("Unable to update jobs on host {} due to exception", host, e);
+                    unableToUpdateJobCounter.increment();
+                    result = false;
                 }
-            );
+            } else {
+                result = false;
+            }
+            return result;
+        });
+        log.info("Finished checking for cluster node health.");
+    }
 
-        // Increment or add new bad nodes
-        badNodes.forEach(
-            host -> {
-                if (this.errorCounts.containsKey(host)) {
-                    this.errorCounts.put(host, this.errorCounts.get(host) + 1);
-                } else {
-                    this.errorCounts.put(host, 1);
+    private void updateJobsToFailedOnHost(final String host) {
+        final Set<Job> jobs = jobSearchService.getAllActiveJobsOnHost(host);
+        jobs.forEach(
+            job -> {
+                try {
+                    jobPersistenceService.setJobCompletionInformation(
+                        job.getId(),
+                        JobExecution.LOST_EXIT_CODE,
+                        JobStatus.FAILED,
+                        "Genie leader can't reach node running job. Assuming node and job are lost."
+                    );
+                    lostJobsCounter.increment();
+                } catch (final GenieException ge) {
+                    log.error("Unable to update job {} to failed due to exception", job.getId(), ge);
+                    Throwables.propagate(ge);
                 }
             }
         );
+    }
 
-        // Find any hosts that are now healthy since last iteration
-        // Two loops to avoid concurrent modification exception
-        final Set<String> toRemove = this.errorCounts.keySet()
-            .stream()
-            .filter(host -> !badNodes.contains(host))
-            .collect(Collectors.toSet());
-        toRemove.forEach(this.errorCounts::remove);
+    private void validateHostAndUpdateErrorCount(final String host) {
+        final boolean isNodeHealthy = isNodeHealthy(host);
+        //
+        // If node is healthy, remove the entry from the errorCounts.
+        // If node is not healthy, update the entry in errorCounts
+        //
+        if (isNodeHealthy) {
```

Comment:
Just call the method within the if statement, no need to save it to a local variable